Make StateMachine::new consume its LineNumberProgramHeader

fitzgen opened the pull request:
Because executing a line number program can define new file entries, it is
incorrect to pass the same header to multiple `StateMachine`s, which is possible
when the `StateMachine` doesn't consume the header.

Fixes #40

@philipc, mind taking a look?

Unfortunately, this ended up making `OpcodesIter::next_opcode` take a `&header` param instead of storing the reference in the `OpcodesIter` struct itself, but I'm not sure how we could work around that.

This doesn't have any effect on the benches.

philipc commented:
Storing a copy of the header in `LineNumberRow` isn't good. If the user does want to clone the row then that'll be expensive. I think it will be better to store the header in `StateMachine`. The problem with that is `LineNumberRow` can no longer contain a reference to the header (because that would be a self-ref in the `StateMachine`). But `LineNumberRow` doesn't actually need the entire header, just one filename, so we can change `LineNumberRow` to only have a `&'input [u8]` for the filename that it needs. Not sure if this will fix the `OpcodesIter` problem too (probably not).

You should also be able to remove the `RefCell` usage.

philipc commented:
Actually the row needs a `FileEntry`, so storing that isn't as nice.

fitzgen commented:
@philipc ok, I moved the header into `StateMachine`, removed the `RefCell` usage, and I think things got a lot nicer in general. One small hitch is that `LineNumberRow::file` needs to take the header as input now, but I think that's a fine balance (probably need to add a `StateMachine::header` getter, too, so users can call this method).

fitzgen commented:
> (probably need to add a StateMachine::header getter, too, so users can call this method)

Done.

fitzgen commented:
Oh, and this is probably going to conflict with #47 but I can wait and rebase on top of that.

philipc commented:
I like how this is shaping up.

Maybe `LineNumberRow::file` should take the `StateMachine` as input instead of the header so that the user can do `row.file(&state_machine)` instead of `row.file(state_machine.header())`, and also to disallow the incorrect `row.file(&header)`.

fitzgen commented:
> Maybe LineNumberRow::file should take the StateMachine as input instead of the header so that the user can do row.file(&state_machine) instead of row.file(state_machine.header()), and also to disallow the incorrect row.file(&header).

I don't think it is really any more difficult to pass the incorrect `StateMachine` than the incorrect `header`. I toyed with adding phantom data for `'input` and `Endian` to LineNumberRow, but that only stops the user from using headers from completely different object files, not different headers in the same object file which seems much more likely an error. The phantom lifetime also artificially restricts against keeping them in the heap indefinitely.

Do you still think passing the state machine is the way to go?

philipc commented:
> I toyed with adding phantom data for 'input

The phantom data for a lifetime will still only be a bound, not an exact matching lifetime, right?

> Do you still think passing the state machine is the way to go?

I don't think it would be bad to do, but I don't feel strongly about it.

philipc commented:
I looked at what changes would be needed in addr2line for this, and it's not clean. Because `next_row()` is returning a reference, it keeps a mutable borrow for the state machine, so you can't do anything else (like call `state_machine.header()`).

It seems to me that either `next_row()` shouldn't be returning a reference, or the row should be self contained (so you don't need to call additional functions on the state machine while processing it). The former requires copying the row, the latter requires copying the file entry into the row. Or maybe I'm just not good enough at working with the borrow checker yet.

fitzgen commented:
Ok after a bit of juggling I have a design that plays well with the borrow checker! I'm really happy with how this came out.

@philipc want to take a final look?

fitzgen commented on src/line.rs:
```diff
@@ -1160,6 +1169,23 @@ impl<'input> FileEntry<'input> {
+    /// If this file's directory index is 0, meaning that it was found in
+    /// the current directory of compilation, return `None`.
+    pub fn directory<Endian>(&self,
+                             header: &LineNumberProgramHeader<'input, Endian>)
```
I'd like this API to take a UnitHeader or something and attempt to parse out the current directory, too, but I think this can be done in a follow up.

philipc commented:
I think this is good.  I suspect this won't be able to implement `FallibleIterator` now though, due to the lifetime on `next_row`. Providing helper APIs like `run_to_address` can address that though.

fitzgen commented:
> I think this is good. I suspect this won't be able to implement FallibleIterator now though, due to the lifetime on next_row. Providing helper APIs like run_to_address can address that though.

Yeah, that's too bad, but I think it is the right trade off.
